### ISSUE TITLE
`Rune.constant` name

### ISSUE THREAD

**Issue by harrysolovay** [paritytech/capi#569]
Still feels confusing imo, given that there's nothing preventing someone from inputting a computed value. Iirc @tjjfvi doesn't like `Rune.lift` bc it may be fn-specific. I don't know if this is always the case... doesn't seem to be for Zio. Thoughts on alternative naming / whether `Rune.constant` is the most fitting?

**Comment by harrysolovay** [comment]
Yes, it's constant from the perspective of Rune. Still, might pose some confusion to end devs. I suppose the same argument could be made for renaming "map" though, which would be a bad rename imo. @ryanleecode @nythrox thoughts on `Rune.constant` naming?

**Comment by ryanleecode** [comment]
`of`, `lift`, and `from` are the common methods. But im ok `constant`. But my 1st preference `outside` of constant would be `of`

**Comment by nythrox** [comment]
I like `Rune.value` because it's intuitive (for non functional programmers) and familiar for users of `Promise.value` or `Result.value`

**Comment by ryanleecode** [comment]
Constant can be confusing because the value isnt actually "constant"

**Comment by tjjfvi** [comment]
`from` is good for a `FooLike -> Foo` IMO.

`Rune.value` would be fine except for that we have `ValueRune`, and not all `ValueRune`s are `Rune.value`s

**Comment by harrysolovay** [comment]
`ValueRune` isn't a great name either, as all Runes resolve to values of some sort. Perhaps we could use `Rune.value` if we did away with `ValueRune` entirely. In my usage, I must frequently `.into(ValueRune)`... to an extent that I believe the benefit of including its methods in `Rune` would outweigh the precedence/visibility concern of Rune-specific methods in autocompletion menus.

**Comment by tjjfvi** [comment]
Part of the point of `ValueRune` is that it's less privileged; just like every other source of rune methods, you have to `.into` it if it's not already an instance.

Once we're adding methods to `Rune` itself, where do we draw the line? Should `ArrayRune` methods be on `Rune`? `MetaRune`? Right now, it's natural to have them as separate classes, but if `ValueRune` isn't a separate class, perhaps those others should be, either.

The fluent API was markedly less ergonomic when all of the methods were present. `.into(ValueRune)` isn't an unfortunate thing you have to do, it's a good, intentional step that says "I'm leaving the capi fluent API and treating this thing as a plain value".

> `ValueRune` isn't a great name either, as all Runes resolve to values of some sort.

A `FooRune` doesn't just mean "a `Rune` that resolves to `Foo`", it means "a `Rune` that is being *used* as a `Foo`". Just like a `Rune<number[]>` isn't necessarily a `ArrayRune<number>`, a `Rune<number>` isn't necessarily a `ValueRune<number>`.

Sure, all runes resolve to values, but you don't always care about that. When you care about the fact that it's resolving to a value, you use a `ValueRune`. If you instead care about it representing, say, a pallet, you use a `PalletRune`.

**Comment by harrysolovay** [comment]
> it's a good, intentional step that says "I'm leaving the capi fluent API and treating this thing as a plain value"

True. Given that all runes can become `ValueRune`s, would you be opposed to `Rune` having a `value` method so that we can convert to `ValueRune` without needing to import `ValueRune` itself? This could be a nice compromise.

Meanwhile, since not all Runes can become `ArrayRune`s, so this would continue to require explicit `.into`ing. Hopefully this somewhat answers your question:

> where do we draw the line?

If x useful enough and possible for all Runes, it'd be nice to make x accessible without the extra import.

**Comment by tjjfvi** [comment]
> Given that all runes can become `ValueRune`s, would you be opposed to `Rune` having a `value` method so that we can convert to `ValueRune` without needing to import `ValueRune` itself?

This means that there are two different ways to cast to different rune types, which I don't think is good. Having one unified API for manipulating rune types reduced cognitive burden, and it means that `ValueRune` remains unprivileged.

**Comment by harrysolovay** [comment]
Good point. You're right.

Back to the topic of this issue. What are some viable alternatives to the name constant?

**Comment by tjjfvi** [comment]
I don't think we need a better name for constant.

Most users will either not need to use this, or should just use `Rune.resolve`, which as precedent from `Promise`.

`Rune.constant` is mostly needed internally, where `constant` is an apt name.

**Comment by harrysolovay** [comment]
This depends on whether we're going to somehow enforce a convention for turning constants into runes. For instance, when initializing a `MultisigRune`:

```ts
const props = {
  signatories: [alice.publicKey, bob.publicKey, charlie.publicKey],
  threshold: 2,
}
```

We have a static method on `Multisig`, which someone can use as follows:

```ts
const multisig = MultisigRune.from(client, props)
```

However, we cannot be sure that lib devs will follow this convention in their own Runes. I believe this approach is more likely to be consistently followed.

```ts
const multisig = Rune.constant(props).into(MultisigRune)
```

Yes, the developer could instead use `Rune.resolve`, and it isn't terribly costly... but it feels unclean. If anything, maybe we do away with `Rune.constant` entirely?

**Comment by tjjfvi** [comment]
I think `Rune.resolve` should be renamed to `Rune.from`. Then, it can be redesigned to work with subclasses, such that `FooRune.from(runic, ...ctorArgs)` is a shorthand for `Rune.resolve(runic).into(FooRune, ...ctorArgs)` (though I would need to confirm that the typing is fine).

> If anything, maybe we do away with `Rune.constant` entirely?

This isn't viable, because `Rune.constant` is very important in the context of `Rune<Rune<>>`s.
